### DISCUSSION TITLE
What are macros and how do they work? I also have a very low IQ

### DISCUSSION THREAD

**Discussion by ghost** [community/community#109264] [Programming Help]
### Body

Hi,

I am making an Assembler for Atari Jaguar so I can make a remake of Space Invaders and I written the following code:
```c
// 16 bit registers

// AX
#ifndef AX
    #define AX 0x000000000000000
#endif

// BX
#ifndef BX     
    #define BX 0x000000000000000
#endif 

// CX
#ifndef CX
    #define CX 0x000000000000000
#endif 

// SI
#ifndef SI     
    #define SI 0x000000000000000
#endif

// DI
#ifndef DI
    #define DI 0x000000000000000
#endif

// 32 bit registers

// EAX
#ifndef EAX 
    #define EAX 0x000000000000000000000000000000
#endif 

// EBX
#ifndef EBX 
    #define EBX 0x000000000000000000000000000000
#endif 

// ECX
#ifndef ECX 
    #define ECX 0x000000000000000000000000000000
#endif 

// ESI
#ifndef ESI
    #define ESI 0x000000000000000000000000000000
#endif

// EDI
#ifndef EDI 
    #define EDI 0x000000000000000000000000000000
#endif

// Assembly Instructions

// Comment
#ifndef COMMENT
    #define COMMENT "COMMENT"
#endif

// Multi-line Comment
#ifndef MULI_LINE_COMMENT
    #define MULTI_LINE_COMMENT "MULTI_LINE_COMMENT"
#endif

// End Multi-line Comment
#ifdef END_MULTI_LINE_COMMENT
    #define END_MULTI_LINE_COMMENT "END_MULI_LINE_COMMENT"
#endif

// Pointer
#ifndef POINTR
    #define POINTR "POINTER"
#endif 
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by davea42** [comment]
It seems appropriate to assume C/C++ is the relevant language. ?

Compilation of a program is not a monolithic one-shot process.
Instead things proceed in an order defined  by the language standard.

C/C++ specify an pre-compilation step of (referring to your question)
reading the C/C++ source (and ignoring comments) and remembering
applicable #defines and whenever a defined string is encountered,
replacing the define name with the string with its value.

This step is referred to as the 'preprocessor'.

Specifically, if POINTR is not defined when your text is read,
any occurrence of POINTR  is replaced by "POINTER"
and the quotes are also part of the replacement.
The reading and replacements occur in order as presented
by the C/C++ source file.
(if POINTR is  defined when the #ifndef is seen then
whatever provided  the POINTR value is some other string
and  the replacement value is that other string.)
The replacement string can be empty, such as
 
#define NOTDEF

Those replacement strings can be ... anything.  Whether the result of the replacement
is valid C/C++ matters, but is not a concern of the preprocessor!

Once the replacements specified by the preprocessor are done the entire result 
after the entire source file is preprocessed ( including all #includes) is presented 
to the compiler to do the compilation (and the compiler-proper 
never sees a #define or #include or any other preprocessor cmmand)..... or produce error reports.

Each #include means grab the full text of the referred-to header
(if it can be found, else it is an error)  and preprocess that 
as if it appeared in your C/C++ souirce.

At the point of actual compilation all the macro commands (#define #undef #include etc)
have been removed and what is left is C/C++ source without the macro commands.

I hope this is a relevant response and I hope it helps.

**Reply by ghost** [reply]
I am using a C header file, but eventually I am going to use C++ for the compiler.
